AI #1785: Add Invoice Detail and Billing Period datasets (#1900)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/invoice_detail/dataset.md

## Relationships<!--SkipTOC-->

The Invoice Detail dataset can be joined to the [Cost and Usage](#dataset.costandusage) dataset through the use of either Invoice ID or Invoice Detail ID.  Take note: one or both datasets will need to be aggregated in order to facilitate any comparison.  For more information, see the Invoice Reconciliation(#supportedfeatures.invoicereconciliation) supported feature.

| Dataset A      | Dataset A Column   | Dataset B      | Dataset B Column   |
| :------------- | :----------------- | :------------- | :----------------- |
| Invoice Detail | Invoice ID         | Cost and Usage | Invoice ID         |
| Invoice Detail | Invoice Detail ID  | Cost and Usage | Invoice Detail ID  |

## Requirements<!--SkipTOC-->

InvoiceDetail adheres to the following requirements:

* InvoiceDetail MUST be present if the invoice issuer supports payable invoices.
* InvoiceDetail MUST conform to [ColumnHandling](#attributes.columnhandling) requirements.
* InvoiceDetail MUST conform to [NullHandling](#attributes.nullhandling) requirements.
* InvoiceDetail MUST conform to [InvoiceHandling](#attributes.invoicehandling) requirements.

> 

Reviewer: I know chicken and egg but PR (https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1865) would have us move the presence rules to the dataset level like:

```suggestion
The presence of columns in InvoiceDetail MUST adhere to the following requirements:
  * BilledCost MUST include [BilledCost](#datasets.invoicedetail.billedcost).
```

Author: As discussed: if PR #1865 clears TF review in tomorrow's TF-RM meeting and is ready for Member review, I'm happy to revise this PR to use that format, presuming it will be approved by the Members.

Author: PR #1865 has cleared TF review.  Thus, I have applied that pattern of column presence requirements issued against the dataset to this PR.  Resolving.

| [Billing Period End](#datasets.invoicedetail.billingperiodend)                   | Dimension   | Mandatory     | False        | Date/Time |
| [Billing Period Start](#datasets.invoicedetail.billingperiodstart)               | Dimension   | Mandatory     | False        | Date/Time |
| [Invoice Detail Created](#datasets.invoicedetail.invoicedetailcreated)           | Dimension   | Mandatory     | False        | Date/Time |
| [Invoice Detail Description](#datasets.invoicedetail.invoicedetaildescription)   | Dimension   | Mandatory     | True         | String    |
| [Invoice Detail Grain](#datasets.invoicedetail.invoicedetailgrain)               | Dimension   | Mandatory     | True         | JSON      |
| [Invoice Detail ID](#datasets.invoicedetail.invoicedetailid)                     | Dimension   | Mandatory     | False        | String    |
| [Invoice Detail Last Updated](#datasets.invoicedetail.invoicedetaillastupdated)   | Dimension   | Mandatory     | False        | Date/Time |
| [Invoice ID](#datasets.invoicedetail.invoiceid)                                 | Dimension   | Mandatory     | False        | String    |

> | [Invoice Issue Date](#datasets.invoicedetail.invoiceissuedate)                   | Dimension   | Mandatory     | False        | Date/Time |

Reviewer 2: Shifting this to TRUE based on the invoiceissuedate.md requirement that states that InvoiceIssueDate MAY be null

```suggestion
| [Invoice Issue Date](#datasets.invoicedetail.invoiceissuedate)                   | Dimension   | Mandatory     | True        | Date/Time |
```

| [Invoice Status](#datasets.invoicedetail.invoicestatus)                         | Dimension   | Mandatory     | False        | String    |
| [Payment Currency](#datasets.invoicedetail.paymentcurrency)                     | Dimension   | Mandatory     | False        | String    |
| [Payment Currency Billed Cost](#datasets.invoicedetail.paymentcurrencybilledcost) | Metric      | Conditional   | False        | Decimal   |
| [Payment Currency Invoice Detail ID](#datasets.invoicedetail.paymentcurrencyinvoicedetailid) | Dimension | Conditional | False | String |
| [Payment Due Date](#datasets.invoicedetail.invoicepaymentduedate)               | Dimension   | Mandatory     | True         | Date/Time |
| [Payment Terms](#datasets.invoicedetail.invoicepaymentterms)                     | Dimension   | Mandatory     | False        | String    |
| [Purchase Order Number](#datasets.invoicedetail.purchaseordernumber)             | Dimension   | Conditional   | False        | String    |
| [Reference Invoice ID](#datasets.invoicedetail.referenceinvoiceid)               | Dimension   | Mandatory     | False        | String    |

## Relationships

> The Invoice Detail dataset can be joined to the [Cost and Usage](#datasets.costandusage) dataset through the use of either Invoice ID or Invoice Detail ID. Take note: one or both datasets will need to be aggregated in order to facilitate any comparison. For more information, see the [Invoice Reconciliation](#supportedfeatures.invoicereconciliation) supported feature.

Reviewer 2: I think it's worth clarifying the expected behavior for C&U records in open billing periods where no invoice detail exists yet in this section as well.

Author: I'm not following.  Can you please leave a suggestion?

Reviewer 2: To clarify my original comment: the timing of when InvoiceId and InvoiceDetailId are populated varies across data generators. Some providers (e.g., OCI) populate InvoiceId while the billing period is still open, while others (e.g., AWS) do not publish an InvoiceId until after the billing period is closed. I think this variance is worth calling out so practitioners know what to expect. Suggested addition after the Relationships paragraph, though it may require additional revision:

```suggestion
The availability of InvoiceId and InvoiceDetailId in Cost and Usage datasets vary across data generators. Some data generators populate these values while the billing period is still open, while others do not populate them until after the billing period is closed and invoices have been issued.
```

Author: Okay.  How about this:

```suggestion
The Invoice Detail dataset can be joined to the [Cost and Usage](#datasets.costandusage) dataset through the use of either Invoice ID or Invoice Detail ID. Take note: one or both datasets will need to be aggregated in order to facilitate any comparison. 

The timing of Invoice ID and Invoice Detail ID availability in Cost and Usage varies across data generators. Some data generators populate these values while the [*billing period*](#glossary:billing-period) is still open, while others do not populate them until after the *billing period* is closed and invoices have been issued.

For more information, see the [Invoice Reconciliation](#supportedfeatures.invoicereconciliation) supported feature.
```

| Dataset A      | Dataset A Column   | Dataset B      | Dataset B Column   |
| :------------- | :----------------- | :------------- | :----------------- |
| Invoice Detail | Invoice ID         | Cost and Usage | Invoice ID         |
| Invoice Detail | Invoice Detail ID  | Cost and Usage | Invoice Detail ID  |

> ## Requirements

Reviewer 3: Adding a couple of InvoiceDetail Requirement Candidates (originally specified as InvoiceHandling version 1.4 attribute requirements)

> * InvoiceDetail *FOCUS dataset* MUST account for all monetary line items included on any invoice issued to a BillingAccountId.
> * InvoiceDetail *FOCUS dataset* MAY omit informational line items with zero monetary impact included on invoice only for transparency (e.g., tax exemption notifications, SLA credit details when the credit is already applied to the charged amount).
> * InvoiceDetail *FOCUS dataset* MUST have its invoice line items representation documented and accessible to practitioners, specifying how InvoiceDetail records correspond to individual invoice line items, and indicating whether any informational items with zero monetary impact are excluded from the dataset.
> * InvoiceDetail *FOCUS dataset* MUST have its invoice reconciliation process documented and accessible to practitioners, including a list of columns from both CostAndUsage and InvoiceDetail FOCUS datasets used in reconciliation.

And one CostAndUsage Requirement Candidate (Note: this particular requirement must be rephrased, i.e. tied to InvoiceDetailGrain as well)

> * CostAndUsage *FOCUS dataset* MUST include Custom columns (e.g., x_ChargeSubType) needed to support invoice reconciliation when FOCUS columns are not sufficient.

---

I hope that, when combined with the InvoiceDetail.BilledCost requirements, these requirements cover the InvoiceHandling requirements specified in version 1.3.

@Matt-Cowsert thoughts?

---

Attaching InvoiceHandling v.1.3 Requirements and Exceptions for context.

> ### Requirements
> 
> * All costs that appear on any invoice issued to a [*BillingAccountId*](#datasets.costandusage.billingaccountid) MUST be included in the *FOCUS dataset*.
> * If an invoice-level *charge* appears on a customer invoice but cannot be expressed using existing FOCUS columns, data generators MUST include provider-defined columns (e.g., x_ChargeSubType) to capture the non-FOCUS-defined details needed to support invoice *charges* reconciliation using the *FOCUS dataset*.
> 
> ### Exceptions
> 
> * Informational line items that have zero monetary impact and are included solely for transparency MAY be excluded. Examples include:
>   * Tax exemption notifications
>   * SLA credit details when the credit is already applied to the charged amount
> * If such informational items are excluded, data generators MUST document this in their FOCUS implementation guide and ensure the sum of included charges still equals the invoice total.
> 
>

Author: Acknowledged.  I need to take an action to incorporate these into the PR, as agreed.

Reviewer 3: Additionally, copying over the conversation about one of these requirements originally discussed in PR #1501 (see [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501/files/765c06a04a0d0bfc10025c8341ab501f1e06b7ad#r2813584193)) for context.

**Requirement:**

> * CostAndUsage *FOCUS dataset* MUST include Custom columns (e.g., x_ChargeSubType) needed to support invoice reconciliation when FOCUS columns are not sufficient.

**Comment by @shawnalpay :**

> This requirement could either move to the Cost and Usage dataset level, or the Invoice Detail Grain column level within Invoice Detail.
> 
> @mike-finopsorg What do you think? Let's discuss in the next TF-RM.

**Response by @mike-finopsorg:**

> If we believe InvoiceId and InvoiceDetailId is enough for CostAndUsage and the additional data is only needed in the Invoice Detail Grain column. Then moving to Invoice Detail Grain sounds right.
> 
> If data in Invoice Detail Grain column is not enough and we expect that CostAndUsage dataset needs custom columns still, in this situation then a separate rule in the CostAndUsage dataset is needed.

---

Personally, I would prefer that this level of information be represented in both the InvoiceDetailGrain and the CostAndUsage dataset. 

In FOCUS 1.3, via the InvoiceHandling attribute, we already required that custom columns be included in the CostAndUsage dataset. Now we have an opportunity to rephrase it and specify that CostAndUsage columns MUST be aligned with InvoiceDetailGrain properties, rather than aligning directly with invoice line items.

In other words, I am advocating for 
- InvoiceDetailGrain properties to be aligned with invoice line items, 
- and for the CostAndUsage dataset to be aligned with InvoiceDetailGrain properties.

If I am interpreting the InvoiceDetailGrain requirements correctly, this PR already addresses the alignment between InvoiceDetailGrain and invoice line items. What remains is to **refine the wording of the CostAndUsage dataset requirement**, i.e., to ensure alignment between CostAndUsage and InvoiceDetailGrain.

Reviewer 3: #### Note on the use of the term “monetary line items” (see [original thread_1](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501/files/765c06a04a0d0bfc10025c8341ab501f1e06b7ad#r2813620726) and [thread_2](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501#discussion_r2813601103) in PR #1501):

We agreed that “monetary line items” and “line items with no monetary impact” are vague and that we should be using explicit, testable wording instead, i.e.,

- **"...when BilledCost is not 0"**
- **"...when BilledCost is 0"**

Also worth noting: although the InvoiceHandling intro mentions both **invoices** and **usage statements**, the original intent of the requirements was to focus on invoice line items, i.e.:

- Line items with BilledCost <> 0 → MUST be included
- Line items with BilledCost = 0 → MAY be omitted (or SHOULD be included)

**Callout:** in **usage statements** we expect line items with BilledCost = 0, so we should be mindful of that when applying this wording.

---

### 1) rephrase this requirement to be more specific,

InvoiceDetail Requirement Candidates with more explicit wording:

> * InvoiceDetail FOCUS dataset MUST account for all invoice line items with BilledCost other than 0 from any invoice issued to a BillingAccountId.
> * InvoiceDetail FOCUS dataset MAY omit informational invoice line items where BilledCost is 0 that are included on an invoice only for transparency (e.g., tax exemption notifications, SLA credit details when the credit is already applied to the charged amount).
> * InvoiceDetail FOCUS dataset MUST have its invoice line items representation documented and accessible to practitioners, specifying how InvoiceDetail records correspond to individual invoice line items, and indicating whether any line items where BilledCost is 0 are excluded from the dataset.
>* InvoiceDetail FOCUS dataset MUST have its invoice reconciliation process documented and accessible to practitioners, including a list of columns from both CostAndUsage and InvoiceDetail FOCUS datasets used in reconciliation.

### 2) whether to apply this to Cost and Usage, Invoice Detail, or both.

The combination of these InvoiceDetail dataset requirements and the sum check requirements on InvoiceDetail.BilledCost column effectively ensures that the same line items exist in both datasets. Missing or extra line items in either dataset will cause the sum checks to fail.

**Assumption:** Any error returned by the sum check requirement indicates that the requirement is not satisfied.

Author: @ijurica 
I have incorporated the above bullet points into the definition of the InvoiceDetail dataset.

Must we specify the following requirement?  It strikes me as wholly unnecessary.  We are already saying, "you must include all non-zero Billed Cost line items".  This strikes me as, essentially, the opposite.  I therefore propose excluding it.

> InvoiceDetail FOCUS dataset MAY omit informational invoice line items where BilledCost is 0 that are included on an invoice only for transparency (e.g., tax exemption notifications, SLA credit details when the credit is already applied to the charged amount).

Also, the documentation requirements felt overloaded to me, so I have broken them apart into separate bullets.  Give them a look and let me know what you think.

Reviewer 3: LGTM. Resolving

# Invoice Detail

> The Invoice Detail dataset is a transactional dataset that represents the financial record of [*charges*](#glossary:charges) as they appear on invoices provided by an invoice issuer. This dataset enables FinOps practitioners to perform financial reconciliation, tax reporting, and payment processing tasks. While the [Cost and Usage](#datasets.costandusage) dataset provides granular visibility into consumption, the Invoice Detail dataset ensures alignment with the physical or electronic billing documents.

Reviewer 3: ```suggestion
The Invoice Detail dataset is a transactional dataset that represents the financial record of [*charges*](#glossary:charge) as they appear on invoices provided by an invoice issuer. This dataset enables FinOps practitioners to perform financial reconciliation, tax reporting, and payment processing tasks. While the [Cost and Usage](#datasets.costandusage) dataset provides granular visibility into consumption, the Invoice Detail dataset ensures alignment with the physical or electronic billing documents.
```

| [Invoice Issue Date](#datasets.invoicedetail.invoiceissuedate)                   | Dimension   | Mandatory     | True        | Date/Time |
| [Invoice Issue Status](#datasets.invoicedetail.invoiceissuestatus)             | Dimension   | Mandatory     | False        | String    |
| [Invoice Issuer Name](#datasets.invoicedetail.invoiceissuername)                 | Dimension   | Mandatory     | False        | String    |
| [Payment Currency](#datasets.invoicedetail.paymentcurrency)                     | Dimension   | Mandatory     | False        | String    |
| [Payment Currency Billed Cost](#datasets.invoicedetail.paymentcurrencybilledcost) | Metric      | Conditional   | False        | Decimal   |
| [Payment Currency Invoice Detail ID](#datasets.invoicedetail.paymentcurrencyinvoicedetailid) | Dimension | Conditional | False | String |
| [Payment Due Date](#datasets.invoicedetail.paymentduedate)               | Dimension   | Mandatory     | True         | Date/Time |
| [Payment Terms](#datasets.invoicedetail.paymentterms)                     | Dimension   | Mandatory     | False        | String    |

> | [Purchase Order Number](#datasets.invoicedetail.purchaseordernumber)             | Dimension   | Conditional   | False        | String    |

Reviewer 3: ```suggestion
| [Purchase Order Number](#datasets.invoicedetail.purchaseordernumber)             | Dimension   | Conditional   | True        | String    |
```

| [Reference Invoice ID](#datasets.invoicedetail.referenceinvoiceid)               | Dimension   | Mandatory     | False        | String    |

## Relationships<!--SkipTOC-->

> The Invoice Detail dataset can be joined to the [Cost and Usage](#datasets.costandusage) dataset through the use of either Invoice ID or Invoice Detail ID. Take note: one or both datasets will need to be aggregated in order to facilitate any comparison.

Reviewer 3: Should the join key include `InvoiceIssuerName` as well, e.g.:

a) Invoice Issuer Name, Invoice ID and Invoice Detail ID
b) Invoice Issuer Name and Invoice ID

Author: How about this?

```suggestion
The Invoice Detail dataset can be joined to the [Cost and Usage](#datasets.costandusage) dataset through Invoice Issuer Name, Invoice ID, and (optionally) Invoice Detail ID. Take note: one or both datasets will need to be aggregated in order to facilitate any comparison.
```

The timing of Invoice ID and Invoice Detail ID availability in Cost and Usage varies across data generators. Some data generators populate these values while the [*billing period*](#glossary:billing-period) is still open, while others do not populate them until after the *billing period* is closed and invoices have been issued.

For more information, see the [Invoice Reconciliation](#supportedfeatures.invoicereconciliation) supported feature.

| Dataset A      | Dataset A Column   | Dataset B      | Dataset B Column   |
| :------------- | :----------------- | :------------- | :----------------- |
| Invoice Detail | Invoice ID         | Cost and Usage | Invoice ID         |

> | Invoice Detail | Invoice Detail ID  | Cost and Usage | Invoice Detail ID  |

Author: @ijurica Per the above:

```suggestion
| Dataset A      | Dataset A Column   | Dataset B      | Dataset B Column   |
| :------------- | :----------------- | :------------- | :----------------- |
| Invoice Detail | Invoice Issuer Name and Invoice ID         | Cost and Usage | Invoice Issuer Name and Invoice ID         |
| Invoice Detail | Invoice Issuer Name, Invoice ID, and Invoice Detail ID  | Cost and Usage | Invoice Issuer Name, Invoice ID, and Invoice Detail ID  |
```

## Requirements

> InvoiceDetail adheres to the following requirements:

Reviewer 4: ```suggestion
InvoiceDetail MUST adheres to the following requirements:
```

Author: Agree with this one too, but needs a small tweak:

```suggestion
InvoiceDetail MUST adhere to the following requirements:
```

  * InvoiceDetail MUST include [PaymentCurrencyInvoiceDetailId](#datasets.invoicedetail.paymentcurrencyinvoicedetailid) if the invoice issuer represents billing currency and payment currency at different aggregation levels on payable invoices.
  * InvoiceDetail MUST include [PaymentDueDate](#datasets.invoicedetail.paymentduedate).
  * InvoiceDetail MUST include [PaymentTerms](#datasets.invoicedetail.paymentterms).
  * InvoiceDetail MUST include [PurchaseOrderNumber](#datasets.invoicedetail.purchaseordernumber) if the invoice issuer supports customer input of purchase order numbers.
  * InvoiceDetail MUST include [ReferenceInvoiceId](#datasets.invoicedetail.referenceinvoiceid).
* InvoiceDetail MUST conform to [ColumnHandling](#attributes.columnhandling) requirements.
* InvoiceDetail MUST conform to [NullHandling](#attributes.nullhandling) requirements.
* InvoiceDetail MUST conform to [InvoiceHandling](#attributes.invoicehandling) requirements.
* InvoiceDetail MUST represent all invoice line items with a non-zero [BilledCost](#datasets.invoicedetail.billedcost) on any invoice associated with a [BillingAccountId](#datasets.invoicedetail.billingaccountid).

> * InvoiceDetail documentation adheres to the following requirements:

Reviewer 4: ```suggestion
* InvoiceDetail documentation MUST adhere to the following requirements:
```
